File: cs/ch17-04-arithmetic/Money/Bank.cs
using System.Net;

namespace TheSoftwareGorilla.TDD.Money;

public class Bank<T> : ICurrencyConverter<T> where T : ICurrencyHolder<T>
{
    private readonly Dictionary<CurrencyPair, decimal> _rates = new Dictionary<CurrencyPair, decimal>();
    public int RateCount => _rates.Count;

    public T Convert(T holder, string to)
    {
        decimal rate = Rate(holder.Currency, to);
        if (rate == 0)
        {
            throw new InvalidOperationException($"Exchange Rate not found for {holder.Currency} to {to}");
        }
        return holder.NewCurrencyHolder(Math.Round(holder.Amount * rate, 2, MidpointRounding.AwayFromZero), to, this);        
    }

    public decimal Rate(string from, string to)
    {
        if (from.Equals(to))
        {
            return 1;
        }
        decimal rate = _rates.TryGetValue(new CurrencyPair(from, to), out decimal value) ? value : 0;
        return rate;
    }

    public void AddRate(string from, string to, decimal rate)
    {
        if (!_rates.ContainsKey(new CurrencyPair(from, to)))
        {
            _rates.Add(new CurrencyPair(from, to), Math.Round(rate, 8, MidpointRounding.AwayFromZero));
        } 
        else
        {
            _rates[new CurrencyPair(from, to)] = Math.Round(rate, 8, MidpointRounding.AwayFromZero);
        }
        if (!_rates.ContainsKey(new CurrencyPair(to, from)))
        {
            _rates.Add(new CurrencyPair(to, from), Math.Round(1 / rate, 8, MidpointRounding.AwayFromZero));
        }
    }

}

    

File: java/ch17-04-arithmetic/src/main/java/com/thesoftwaregorilla/tdd/money/Bank.java
package com.thesoftwaregorilla.tdd.money;


import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDateTime;
import java.util.HashMap;

public class Bank<T extends ICurrencyHolder<T>> implements ICurrencyConverter<T> {

    private final HashMap<CurrencyPair, BigDecimal> rates = new HashMap<>();

    @Override
    public T convert(T holder, String toCurrency) {
        BigDecimal rate = this.rate(holder.getCurrency(), toCurrency);
        if (rate.equals(BigDecimal.ZERO.setScale(8, RoundingMode.HALF_UP))) {
            throw new ArithmeticException(String.format("Exchange Rate not found for %s to %s", holder.getCurrency(),
                    toCurrency));
        }
        return holder.newCurrencyHolder(holder.getAmount().multiply(rate), toCurrency, this);
    }


    public BigDecimal rate(String from, String to) {
        if (from.equals(to)) {
            return BigDecimal.ONE.setScale(8, RoundingMode.HALF_UP);
        }
        BigDecimal rate = rates.get(new CurrencyPair(from, to));
        return rate != null ? rate : BigDecimal.ZERO.setScale(8, RoundingMode.HALF_UP);
    }

    public void addRate(String from, String to, BigDecimal rate) {
        rates.put(new CurrencyPair(from, to), rate.setScale(8, RoundingMode.HALF_UP));
        if  (!rates.containsKey(new CurrencyPair(to, from))) {
            rates.put(new CurrencyPair(to, from), BigDecimal.ONE.divide(rate, 8, RoundingMode.HALF_UP));
        }
    }
}
